### ISSUE TITLE
CoreCLR fails to initialize when running from a Docker read-only filesystem

### ISSUE THREAD

**Issue by natemcmaster** [dotnet/coreclr#15243]
Originally reported here https://github.com/aspnet/aspnet-docker/issues/331 by @RehanSaeed. I confirmed this issue appears in the microsoft/dotnet base images.

CoreCLR fails to initalize when using the read-only filesystem in Docker.

# Repro
1. Create a Dockerfile with these contents
```dockerfile
FROM microsoft/dotnet:2.0.3-sdk-stretch AS builder
WORKDIR /code/MyApp
RUN dotnet new console && dotnet publish --output /app --configuration Release

FROM microsoft/dotnet:2.0.3-runtime-stretch
COPY --from=builder /app /app
ENTRYPOINT [ "dotnet", "/app/MyApp.dll" ]
```
2. From command-line in the folder containing this dockerfile
```
docker build --tag read-only-test ./
```
3. Launch the container with a read-only filesystem
```
docker run --rm -it --read-only read-only-test
```

# Result
App fails with
```
Failed to initialize CoreCLR, HRESULT: 0x80004005
```

You can gather a trace by executing `docker run --rm -it --read-only -e COREHOST_TRACE=1 read-only-test`
Corehost trace: [log.txt](https://github.com/dotnet/coreclr/files/1510872/log.txt)

From the trace, it appears the initialization reaches this point, and fails when calling `coreclr_initialize`: https://github.com/dotnet/core-setup/blob/v2.0.3/src/corehost/cli/hostpolicy.cpp#L181-L199

# Details

The readonly flag:

> The --read-only flag mounts the container’s root filesystem as read only prohibiting writes to locations other than the specified volumes for the container.

https://docs.docker.com/engine/reference/commandline/run/#mount-volume--v-read-only

`docker info`

```
Containers: 13
 Running: 0
 Paused: 0
 Stopped: 13
Images: 211
Server Version: 17.09.0-ce
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host ipvlan macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 06b9cb35161009dcb7123345749fef02f7cea8e0
runc version: 3f2f8b84a77f73d38244dd690525642a72156c64
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.9.49-moby
Operating System: Alpine Linux v3.5
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 1.934GiB
Name: moby
ID: 7HYI:PRKR:5ASJ:6UAU:SPUF:SA42:PGMK:VAJH:A5JQ:4DCB:XK6U:6CG7
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 18
 Goroutines: 26
 System Time: 2017-11-28T16:58:34.2150442Z
 EventsListeners: 0
Registry: https://index.docker.io/v1/
Experimental: true
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

**Comment by natemcmaster** [comment]
I investigated a little more by tracking system calls with strace. Not sure if this is going down the right path, but figured it might be useful:

Create an image with strace:
```Dockerfile
FROM microsoft/dotnet:2.0.3-sdk-stretch AS builder
WORKDIR /code/MyApp
RUN dotnet new console && dotnet publish --output /app --configuration Release

FROM microsoft/dotnet:2.0.3-runtime-stretch
COPY --from=builder /app /app
RUN apt-get update && apt-get install strace
ENTRYPOINT [ "strace", "-e", "trace=open,close,write", "-y", "dotnet", "/app/MyApp.dll" ]
```
If you run `docker run -it --rm --read-only --cap-add SYS_PTRACE read-only-test`, you'll see the tail of the log is:
```
open("/proc/6/stat", O_RDONLY)          = 10</proc/6/stat>
close(10</proc/6/stat>)                 = 0
open("/proc/6/stat", O_RDONLY)          = 10</proc/6/stat>
close(10</proc/6/stat>)                 = 0
write(2</dev/pts/0>, "Failed to initialize CoreCLR, HR"..., 49Failed to initialize CoreCLR, HRESULT: 0x80004005) = 49
write(2</dev/pts/0>, "\n", 1
)           = 1
+++ exited with 137 +++
```

If you run without `--read-only`,  the app runs successfully. The matching part of its trace shows
```
open("/proc/7/stat", O_RDONLY)          = 10</proc/7/stat>
close(10</proc/7/stat>)                 = 0
open("/proc/7/stat", O_RDONLY)          = 10</proc/7/stat>
close(10</proc/7/stat>)                 = 0
write(12<pipe:[140514]>, "*", 1)        = 1
open("/proc/7/stat", O_RDONLY)          = 11</proc/7/stat>
close(11</proc/7/stat>)                 = 0
```

**Comment by RussKeldorph** [comment]
@janvorli FYI.  Not sure where this lands.  Feel free to change area.

**Comment by janvorli** [comment]
Will get fixed by a fix to #8844

**Comment by mikem8361** [comment]
My PR #16141 is blocked on an arm tizen problem.

**Comment by mikem8361** [comment]
Fixed in #16141.
